src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,100 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ObjectPool.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// Manages a pool of reusable objects. An object retrieved from the pool should be returned when it's no longer being used so that it can be later recycled.
+    /// </summary>
+    /// <typeparam name="T">The type of objects to pool.</typeparam>
+    /// <remarks>This object pool is not thread-safe and not intended to be shared across concurrent requests.
+    /// Consider using the DefaultObjectPool in 8.x once we drop support for netstandard 1.1
+    /// </remarks>
+    internal class ObjectPool<T> where T : class
+    {
+        private readonly Func<T> objectGenerator;
+        private ObjectWrapper[] items;
+        private T firstItem;
+
+        // The pool size is equal to the level of nesting in the response.
+        // 32 is an arbitrary figure. Could be adjusted appropriately.
+        private const int POOL_SIZE = 8;

```

Comment:
Given how generic this class already is, it might make sense for this to be configurable through the constructor instead.

Reply:
Fixed